### DISCUSSION TITLE
Fluid Tank upgrade

### DISCUSSION THREAD

**Discussion by Jvs34** [P3pp3rF1y/SophisticatedBackpacks#89] [Feature / Idea Archive] [answered]
A backpack upgrade that allows you to store ten buckets of one type of fluid at a time through its GUI and with other mods' pipes mods when the backpack is placed in the world.

The advanced version would allow you to set fluid filters and increase the capacity through either through some crafting methods or by placing a stack of buckets in one of its GUI slots.

The automation aspect of the advanced version could be to automatically store the fluid from the bucket in your hand if the upgrade is set-up to do it.

I would also suggest about it trying to fill fluid into items in your inventory that can accept it, (such as another bucket or immersive engineering's drill), but I'm not sure if there's forge capabilities for that kind of stuff.

**Comment by P3pp3rF1y** [comment]
I have a lot of these things in longer term notes and I am definitely open to ideas on how this should work
- basically the thinking is that the upgrade would take part of the inventory slots (my thinking is 3 columns)
- you could have up to 3/4 of these upgrades in the backpack each with separate tank which can have different fluid
- the fluid fill level displays in place of those slots and so do slots to fill the tank / fill fluid containers 
- would scale based on the backpack tier so basic may be something like 8 buckets, iron 16, ...
- stack upgrades could be used to increase the capacity even further, probably scale much faster than with these given how some mods are cheap about infinite tank capacity
- and yes related refill / pump upgrades which will fill bucket in hand / take fluid form the said bucket

**Reply by P3pp3rF1y** [reply]
Also some way of rendering that the bag has fluids in it would be nice, but there I will definitely need to rely on someone who actually has more artistic skill than me.

**Reply by Jvs34** [reply]
I really like the idea of fluids showing up in item slots instead of the separate upgrade tab (that seems like it would make bucketing stuff a simple click just like in refined storage's fluid grid), but I'm not sure if the capacity should scale with backpack tier.

Crafting stack upgrades does seem like a good compromise for increasing total fluid capacity, since when you upgrade a backpack you're already inherently increasing the capacity.

> Also some way of rendering that the bag has fluids in it would be nice, but there I will definitely need to rely on someone who actually has more artistic skill than me.

For that you could always say screw it and go the traveler's backpack way, a small canister on each side of the backpack if you have more than one fluid tank, then you can divide the vertical space inside them to show the various fluid types stacked on top of each other.

**Reply by P3pp3rF1y** [reply]
Actually to be clear what I meant by the tank taking place of slots is that it literally removes slots and places a tank in the gui where they were instead.
That means that with basic tank you would get one tank and 9 slots taken away, with iron backpack you still have just that one tank but now 18 slots taken away, that's why I believe it should scale with the backpack tier as well.
So a bacpack with a single fluid upgrade would have:
basic tier - 18 inventory slots + tank with capacity of 8 buckets
iron - 36 + 16 buckets
gold - 54 + 24 buckets
diamond - 81 + 24 buckets (diamond has same number of rows, but more columns, which means the tank takes away same number of slots in diamond as it does in gold one, but may adjust this to add a bit of capacity)
netherite - 90 + 27 buckets (not sure about scaling this one as it only adds one row of inventory slots)

Of course adding more fluid upgrades would take away more slots (3 columns per upgrade) and replace the with additional tanks capped at 3 upgrades for gold and lower and 4 upgrades for diamond/netherite

**Reply by Jvs34** [reply]
Oh yeah I understand now, sorry.

**Reply by P3pp3rF1y** [reply]
Hey, no need to apologize, I wasn't clear with my explanation in the first place.

**Comment by TheDeviantCrafter** [comment]
As a more vanilla-themed alternative, you could require the player to load up their backpack with buckets. Holding the backpack and right clicking would fill or empty those buckets.

In larger modpacks it could interact with portable tanks in the same way. Since dense liquid storage is already the norm in most packs I don't see any harm in supporting them.

**Reply by P3pp3rF1y** [reply]
here I will definitely have a separate tank as explained above I like that more than trying to handle fluids in many different items and it has more possibilities for additional things that can be added related to it.

**Comment by P3pp3rF1y** [comment, marked as answer]
And the tank upgrade now exists:
- replaces part of slots
- has a tab for input / output
- but allows to simply click bucket on the tank gui to quickly fill/drain
- renders
- works with both inception and stack upgrade

**Reply by Jvs34** [reply]
The upgrade is pretty good however I don't see a way to automatically take the bucket from your hand and drain it into the tank at the moment, is that gonna come with the advanced upgrade or am I missing something?
